### PR TITLE
[OP] Norm option for SoftmaxOutput

### PR DESCRIPTION
**Pull request by antinucleon**
@piiswrong @pluskid 
It is following Caffe's normalization way. However I am not quite sure about Caffe's blob:
https://github.com/rbgirshick/caffe-fast-rcnn/blob/0dcd397b29507b8314e252e850518c5695efbb83/src/caffe/layers/softmax_loss_layer.cpp#L145

For example, is this line ```top[0]->cpu_diff()[0]``` used for getting the ```out_grad[0].shape_[0]```?

Please help me check it then I will fix it. This issue is known that will influence RPN converge.

### CODE REVIEW

**Comment by piiswrong** [comment]
No, it's for getting the loss weight

File: src/operator/softmax_output-inl.h [hunk 1]
```diff
@@ -118,6 +153,28 @@ class SoftmaxOutputOp : public Operator {
...
+        grad /= static_cast<float>(grad.size(0));
+      } else if (param_.normalization == softmaxout_enum::kValid) {
+        int ignore = static_cast<int>(param_.ignore_label);
+        int64_t valid_cnt = 0;
+        float scale = 1.0f;
+        Tensor<cpu, 1, DType> workspace =
+          ctx.requested[softmaxout_enum::kTempSpace].get_space_typed<cpu, 1, DType>(label.shape_,
+                                                                                    NULL);
+        Copy(workspace, label, label.stream_);
+        for (index_t i = 0; i < workspace.size(0); ++i) {
```
Comment (piiswrong) [inline src/operator/softmax_output-inl.h]: doesn't really make sense to use ignore for single output since you can just drop this instance alltogether

File: src/operator/softmax_output-inl.h [hunk 2]
```diff
@@ -105,7 +116,31 @@ class SoftmaxOutputOp : public Operator {
...
+        grad /= static_cast<float>(grad.size(0));
+      } else if (param_.normalization == softmaxout_enum::kValid) {
+        int ignore = static_cast<int>(param_.ignore_label);
+        int64_t valid_cnt = 0;
+        float scale = 1.0f;
+        Tensor<cpu, 2, DType> workspace =
+          ctx.requested[softmaxout_enum::kTempSpace].get_space_typed<cpu, 2, DType>(label.shape_,
+                                                                                    NULL);
+        Copy(workspace, label, label.stream_);
+        for (index_t i = 0; i < workspace.size(0); ++i) {
```
Comment (piiswrong) [inline src/operator/softmax_output-inl.h]: use mshadow reduce. This doesn't work for gpu

File: src/operator/softmax_output-inl.h [hunk 3]
```diff
@@ -105,7 +116,31 @@ class SoftmaxOutputOp : public Operator {
       } else {
           SoftmaxGrad(grad, out, label);
       }
-      grad *= DType(param_.grad_scale/s3[2]);
+      grad *= DType(param_.grad_scale / s3[2]);
```
Comment (piiswrong) [inline src/operator/softmax_output-inl.h]: combine this with other normalizations to save computation

File: src/operator/softmax_output-inl.h [hunk 4]
```diff
@@ -43,6 +46,14 @@ struct SoftmaxOutputParam : public dmlc::Parameter<SoftmaxOutputParam> {
     DMLC_DECLARE_FIELD(use_ignore).set_default(false)
     .describe("If set to true, the ignore_label value will not contribute "
       "to the backward gradient");
+    DMLC_DECLARE_FIELD(normalization)
```
Comment (piiswrong) [inline src/operator/softmax_output-inl.h]: I don't think this is necessary since 'valid' should automatically be used when using ignore while 'batch' is always done by optimizer.

File: src/operator/softmax_output-inl.h (hunk 1 above)
Comment (pluskid) [inline src/operator/softmax_output-inl.h]: @piiswrong it is used in training with sequences with different lengths. This option is used to drop padded frames.

**Comment by antinucleon** [comment]
I am adding get CPU resource function. The purpose is to make behavior uniform to other toolkit and make it easier to reproduce result.

**Comment by antinucleon** [comment]
Close for separate PR
